### PR TITLE
Fix #4368: Avoid stackoverflows for some cyclic definitions

### LINKED ISSUE #4368: StackOverflow in typer
```scala
trait X {
  type A = B
  type B
}
trait Y {
  type A
  type B = A
}
trait Z extends X with Y
```

```scala
Exception in thread "main" java.lang.StackOverflowError
	at java.lang.invoke.MemberName.getInvocationType(MemberName.java:164)
	at java.lang.invoke.DirectMethodHandle.preparedLambdaForm(DirectMethodHandle.java:167)
	at java.lang.invoke.DirectMethodHandle.make(DirectMethodHandle.java:89)
	at java.lang.invoke.MethodHandles$Lookup.getDirectMethodCommon(MethodHandles.java:1660)
	at java.lang.invoke.MethodHandles$Lookup.getDirectMethodNoSecurityManager(MethodHandles.java:1617)
	at java.lang.invoke.MethodHandles$Lookup.getDirectMethodForConstant(MethodHandles.java:1802)
	at java.lang.invoke.MethodHandles$Lookup.linkMethodHandleConstant(MethodHandles.java:1751)
	at java.lang.invoke.MethodHandleNatives.linkMethodHandleConstant(MethodHandleNatives.java:477)
	at dotty.tools.dotc.core.Types$Type.findMember(Types.scala:661)
	at dotty.tools.dotc.core.Types$Type.memberExcluding(Types.scala:491)
	at dotty.tools.dotc.core.Types$Type.$anonfun$nonPrivateMember$1(Types.scala:482)
	at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
	at dotty.tools.dotc.core.Types$Type.nonPrivateMember(Types.scala:482)
	at dotty.tools.dotc.core.Types$NamedType.memberDenot(Types.scala:1777)
	at dotty.tools.dotc.core.Types$NamedType.reload$1(Types.scala:2025)
	at dotty.tools.dotc.core.Types$NamedType.withPrefix(Types.scala:2037)
	at dotty.tools.dotc.core.Types$NamedType.derivedSelect(Types.scala:1997)
	at dotty.tools.dotc.core.Types$TypeMap.derivedSelect(Types.scala:3902)
	at dotty.tools.dotc.core.Types$ApproximatingTypeMap.derivedSelect(Types.scala:4174)
	at dotty.tools.dotc.core.TypeOps$AsSeenFromMap.$anonfun$apply$4(TypeOps.scala:61)
	at dotty.tools.dotc.reporting.trace$.op1$1(trace.scala:18)
	at dotty.tools.dotc.reporting.trace$.conditionally(trace.scala:20)
	at dotty.tools.dotc.core.TypeOps$AsSeenFromMap.apply(TypeOps.scala:54)
	at dotty.tools.dotc.core.Types$TypeMap.$anonfun$mapOver$7(Types.scala:3974)
	at dotty.tools.dotc.core.Types$VariantTraversal.atVariance(Types.scala:3888)
	at dotty.tools.dotc.core.Types$TypeMap.mapOver(Types.scala:3974)
	at dotty.tools.dotc.core.TypeOps$AsSeenFromMap.$anonfun$apply$4(TypeOps.scala:67)
	at dotty.tools.dotc.reporting.trace$.op1$1(trace.scala:18)
	at dotty.tools.dotc.reporting.trace$.conditionally(trace.scala:20)
	at dotty.tools.dotc.core.TypeOps$AsSeenFromMap.apply(TypeOps.scala:54)
	at dotty.tools.dotc.core.TypeOps.asSeenFrom(TypeOps.scala:28)
	at dotty.tools.dotc.core.TypeOps.asSeenFrom$(TypeOps.scala:27)
	at dotty.tools.dotc.core.Contexts$Context.asSeenFrom(Contexts.scala:71)
	at dotty.tools.dotc.core.Types$Type.$anonfun$asSeenFrom$1(Types.scala:767)
	at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
	at dotty.tools.dotc.core.Types$Type.asSeenFrom(Types.scala:766)
	at dotty.tools.dotc.core.Denotations$SingleDenotation.computeAsSeenFrom(Denotations.scala:1075)
	at dotty.tools.dotc.core.Denotations$SingleDenotation.computeAsSeenFrom(Denotations.scala:660)
	at dotty.tools.dotc.core.Denotations$PreDenotation.asSeenFrom(Denotations.scala:131)
	at dotty.tools.dotc.core.Denotations$SingleDenotation.mapInherited(Denotations.scala:1062)
	at dotty.tools.dotc.core.Denotations$SingleDenotation.mapInherited(Denotations.scala:660)
	at dotty.tools.dotc.core.SymDenotations$ClassDenotation.collect$1(SymDenotations.scala:1605)
	at dotty.tools.dotc.core.SymDenotations$ClassDenotation.collect$1(SymDenotations.scala:1600)
	at dotty.tools.dotc.core.SymDenotations$ClassDenotation.collect$1(SymDenotations.scala:1600)
	at dotty.tools.dotc.core.SymDenotations$ClassDenotation.$anonfun$computeNPMembersNamed$1(SymDenotations.scala:1613)
	at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
	at dotty.tools.dotc.core.SymDenotations$ClassDenotation.computeNPMembersNamed(SymDenotations.scala:1592)
	at dotty.tools.dotc.core.SymDenotations$ClassDenotation.nonPrivateMembersNamed(SymDenotations.scala:1582)
	at dotty.tools.dotc.core.SymDenotations$ClassDenotation.findMember(SymDenotations.scala:1617)
	at dotty.tools.dotc.core.Types$Type.go$1(Types.scala:509)
	at dotty.tools.dotc.core.Types$Type.goThis$1(Types.scala:613)
	at dotty.tools.dotc.core.Types$Type.go$1(Types.scala:522)
	at dotty.tools.dotc.core.Types$Type.findMember(Types.scala:658)
	at dotty.tools.dotc.core.Types$Type.memberExcluding(Types.scala:491)
	at dotty.tools.dotc.core.Types$Type.$anonfun$nonPrivateMember$1(Types.scala:482)
	at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
	at dotty.tools.dotc.core.Types$Type.nonPrivateMember(Types.scala:482)
	at dotty.tools.dotc.core.Types$NamedType.memberDenot(Types.scala:1777)
	at dotty.tools.dotc.core.Types$NamedType.reload$1(Types.scala:2025)
	at dotty.tools.dotc.core.Types$NamedType.withPrefix(Types.scala:2037)
	at dotty.tools.dotc.core.Types$NamedType.derivedSelect(Types.scala:1997)
	at dotty.tools.dotc.core.Types$TypeMap.derivedSelect(Types.scala:3902)
	at dotty.tools.dotc.core.Types$ApproximatingTypeMap.derivedSelect(Types.scala:4174)
	at dotty.tools.dotc.core.TypeOps$AsSeenFromMap.$anonfun$apply$4(TypeOps.scala:61)
	at dotty.tools.dotc.reporting.trace$.op1$1(trace.scala:18)
	at dotty.tools.dotc.reporting.trace$.conditionally(trace.scala:20)
	at dotty.tools.dotc.core.TypeOps$AsSeenFromMap.apply(TypeOps.scala:54)
	at dotty.tools.dotc.core.Types$TypeMap.$anonfun$mapOver$7(Types.scala:3974)
	at dotty.tools.dotc.core.Types$VariantTraversal.atVariance(Types.scala:3888)
	at dotty.tools.dotc.core.Types$TypeMap.mapOver(Types.scala:3974)
	at dotty.tools.dotc.core.TypeOps$AsSeenFromMap.$anonfun$apply$4(TypeOps.scala:67)
	at dotty.tools.dotc.reporting.trace$.op1$1(trace.scala:18)
	at dotty.tools.dotc.reporting.trace$.conditionally(trace.scala:20)
.......
```

### PR DESCRIPTION
**Pull request by odersky**
We use a two-pronged approach:

First, we now detect some cycles that span inherited or intersected classes or traits. Unfortunately, this operation forces the types of type members, which means it can lead to spurious cycles itself. So we can do this only after a lot of types are already computed, which means some cycles lead to stackoverflows before they are detected.

Hence the second strategy: give more useful information when deep recursions happen.

### CODE REVIEW

**Comment by odersky** [comment]
test performance please

File: compiler/src/dotty/tools/dotc/core/Types.scala [hunk 1]
```diff
@@ -647,19 +645,14 @@ object Types {
...
         ctx.pendingMemberSearches = name :: ctx.pendingMemberSearches
-      }
       ctx.findMemberCount = recCount + 1
-      //assert(ctx.findMemberCount < 20)
       try go(this)
       catch {
         case ex: Throwable =>
-          core.println(i"findMember exception for $this member $name, pre = $pre")
-          throw ex // DEBUG
+          core.println(s"findMember exception for $this member $name, pre = $pre, recCount = $recCount")
```
Comment (Blaisorblade) [inline compiler/src/dotty/tools/dotc/core/Types.scala]: Nit: why change formatter to `s`?

**Comment by odersky** [comment]
Yes, #318 is handled, albeit crudely.

File: compiler/src/dotty/tools/dotc/core/Types.scala (hunk 1 above)
Comment (odersky) [reply compiler/src/dotty/tools/dotc/core/Types.scala]: Avoid possible crashes when trying to show `pre`.

**Comment by odersky** [comment]
test performance please

**Comment by odersky** [comment]
So, this is a rather sneaky way to fix a whole class of problems that cause the compiler to stackoverflow in some way or another. One would hope there was a principled way to detect and repair all illegal cyclic references before they cause stackoverflows. Unfortunately, that looks impossible or at least impossibly hard: Scala's type system subsumes F_sub and is therefore undecidable. So some infinite loops/stackoverflows are unavoidable. What's more, checking for cycles early has the unfortunate consequence that we force types and thereby can cause spurious cyclic references! So the mere act of checking generates spurious errors. This is a real problem in practice, observed in both the compiler codebase and stdlib. Another possibility would be to attach a cycle detection component to every operation that might stackoverflow. But these tend to be also operations that are most time-critical, so doing the extra work is prohibitive.

The idea in this pull request is to act only if there _is_ a stackoverflow. In that case we unroll the stack until we hit a safepoint (typically: a point where we typecheck an AST node and therefore have an error position to record). While unrolling we keep a count of all important method frames that were unrolled. Important means: the method contains a try/catch that does the unrolling and counting. At the safepoint, we check which method appeared most often in the unrolled frames. We then compose a nice error message showing only calls info for that method and leaving out the middle of the stack if it is too large. A typical error message looks like this:

    |  val b: y.B = a 
    |         ^
    |         Recursion limit exceeded.
    |         Maybe there is an illegal cyclic reference?
    |         A recurring operation is (inner to outer):
    |         
    |           find-member y.B
    |           find-member y.A
    |           find-member y.B
    |           find-member y.A
    |           find-member y.B
    |           find-member y.A
    |           find-member y.B
    |           find-member y.A
    |           find-member y.B
    |           find-member y.A
    |           ...
    |         
    |           find-member y.A
    |           find-member y.B
    |           find-member y.A
    |           find-member y.B
    |           find-member y.A
    |           find-member y.B
    |           find-member y.A
    |           find-member y.B
    |           find-member y.A
    |           find-member y.B

Of course, it's still nicer to pinpoint the cycle where it arises, and we try to do this also, as long as it can be done safely and efficiently. But we have transformed the game from a bug that needs fixing, (a stackoverflow is a compiler crash, so it's wrong and embarrassing) to a quality of error message question. Such a question can be answered on a case-by-case basis, taking into account the frequency of the error (in my experience most of these cyclic reference errors are synthetic and don't tend to occur in practice) and the effort needed to get a more specific error message.

**Comment by odersky** [comment]
test performance please

**Comment by odersky** [comment]
This PR fixes #4368, #4369, #4370, #4371, #4372 and #318.

**Comment by odersky** [comment]
test performance please

**Comment by odersky** [comment]
I won't invest more time in this. So, ready to get this in?

**Comment by Blaisorblade** [comment]
Hm. Especially since `RecursionOverflow` appears in the error, it seems worth renaming that (and friends) as `RecursionOverflowException`.

But otherwise, it seems reasonable to classify those as requests for enhancements. I'll have to look at the first more closely, because the output contradicts quite a few of my expectations and one of the two must be corrected.

**Comment by smarter** [comment]
I think it'd be good to have a -Y setting to turn off the exception catching, so that the root cause of issues can be investigated more easily.

**Comment by odersky** [comment]
@smarter The recommended way to find out more is to turn the `cyclic` printer on.

**Comment by smarter** [comment]
@odersky The issue is that by turning everything into errors, @alexknvl fuzzer will never detect any stackoverflow again, even though some of them might be legitimate bugs we can fix.

**Comment by Blaisorblade** [comment]
Status: I'd like to wait for #4511 before enabling this by default, lest this PR swallows errors. But that involves reviewing positions in quite a few scenarios.

Given the amount of code and potential for divergence, it could be best to make this conditional as planned, default to false and merge this PR, then flip the default in #4511.

**Comment by Blaisorblade** [comment]
@smarter: can you take a look? This can now be disabled, and the remaining issues here (in comments above) don't seem blockers, so I think this should be mergeable.

**Comment by Blaisorblade** [comment]
Or @odersky can you review my changes? Whoever's faster.

File: compiler/src/dotty/tools/dotc/config/ScalaSettings.scala [hunk 2]
```diff
@@ -146,6 +146,7 @@ class ScalaSettings extends Settings.SettingGroup {
   val Xlink = BooleanSetting("-Xlink", "Recompile library code with the application.")
   val YoptPhases = PhasesSetting("-Yopt-phases", "Restrict the optimisation phases to execute under -optimise.")
   val YoptFuel = IntSetting("-Yopt-fuel", "Maximum number of optimisations performed under -optimise.", -1)
+  val YnoDecodeStacktraces = BooleanSetting("-Yno-decode-stacktraces", "Show operations that triggered StackOverflows instead of printing the raw stacktraces.")
```
Comment (allanrenucci) [inline compiler/src/dotty/tools/dotc/config/ScalaSettings.scala]: Comment is misleading. How I understand it: "if you pass the flag, you won't see original stack traces". I guess it is the opposite

Comment (Blaisorblade) [reply compiler/src/dotty/tools/dotc/config/ScalaSettings.scala]: Oh right, fixing it

**Pull request merged**